program/src/instruction.rs

```diff
@@ -1,184 +1,373 @@
+        cooldown_time_seconds: u64,
         max_deactivation_basis_points: u16,
     },
 
     /// Initializes stake account data for a validator.
-    ///
-    /// NOTE: Anybody can create the stake account for a validator. For new
-    /// accounts, the authority is initialized to the validator vote account's
-    /// withdraw authority.
-    ///
-    /// 0. `[]` Stake config account
-    /// 1. `[w]` Validator stake account
-    ///     * PDA seeds: ['stake', validator, config_account]
-    /// 2. `[]` Validator vote account
-    /// 3. `[]` System program
+    #[account(
+        0,
+        name = "config",
+        desc = "Stake config account"
+    )]
+    #[account(
+        1,
+        writable,
+        name = "stake",
+        desc = "Validator stake account (pda of `['stake', config, validator]`)"
+    )]
+    #[account(
+        2,
+        name = "validator_vote",
+        desc = "Validator vote account"
+    )]
+    #[account(
+        3,
+        name = "system_program",
+        desc = "System program account"
+    )]
     InitializeStake,
 
     /// Stakes tokens with the given config.
-    ///
-    /// Limited to the current amount of SOL staked to the validator.
-    ///
-    /// NOTE: Anybody can stake tokens to a validator, but this does not work
-    /// like native staking, because the validator can take control of staked
-    /// tokens by deactivating and withdrawing.
-    ///
-    /// 0. `[w]` Stake config account
-    /// 1. `[w]` Validator stake account
-    ///     * PDA seeds: ['stake', validator, config_account]
-    /// 2. `[w]` Token Account
-    /// 3. `[s]` Owner or delegate of the token account
-    /// 4. `[]` Validator vote account
-    /// 3. `[]` Stake Token Mint
-    /// 4. `[]` Stake Token Vault, to hold all staked tokens.
-    ///   Must be the token account on the stake config account
-    /// 5. `[]` Token program
-    /// 6.. Extra accounts required for the transfer hook
-    ///
-    /// Instruction data: amount of tokens to stake, as a little-endian u64
+    #[account(
+        0,
+        writable,
+        name = "config",
+        desc = "Stake config account"
+    )]
+    #[account(
+        1,
+        writable,
+        name = "stake",
+        desc = "Validator stake account"
+    )]
+    #[account(
+        2,
+        writable,
+        name = "source_token",
+        desc = "Token account"
+    )]
+    #[account(
+        3,
+        signer,
+        name = "token_authority",
+        desc = "Owner or delegate of the token account"
+    )]
+    #[account(
+        4,
+        name = "validator_vote",
+        desc = "Validator vote account"
+    )]
+    #[account(
+        5,
+        name = "mint",
+        desc = "Stake Token Mint"
+    )]
+    #[account(
+        6,
+        writable,
+        name = "vault_token",
+        desc = "Stake token Vault"
+    )]
+    #[account(
+        7,
+        name = "spl_token_program",
+        desc = "SPL Token 2022 program"
+    )]
```

Comment:
I'm not totally sure on this, but we might not want to make this specifically Token-2022. It could just be "a token program", in case people wanted to use other token programs in the future.

Reply:
Hmmm, how would that work? Like, we need the transfer hook for this to work, right? So if we keep it generic than it "looks" like you could use SPL token.

Reply:
Yeah you're right we do need the extensions, but typically we've been treating Token-2022's functionality as an "interface", just like Token-2022 implements (and extends) the SPL Token interface.

It's not really a perfect pattern right now, but I'd say let's just go with "Token program". The instruction should be compatible with any token program that implements, say "Transfer Hook".